tests/controller/README.md

```diff
@@ -1 +1,209 @@
-This folder contains steps to deploy the controller framework with an example manifest.yaml
\ No newline at end of file
+##What is a controller?
+
+In simple terms, controller is a framework to keep the registry up-to-date. The idea for this framework is inspired from the [Kubernetes controller](https://kubernetes.io/docs/concepts/architecture/controller/) pattern. Controller can be used to enrich the API data stored in the registry, with additional analytical artifacts. 
+
+### How does it work?
+* A Controller will take in configuration (aka manifest) which describes the desired state of the registry. The desired state in this context is essentially dependency relations between different entities in the database. 
+* The controller’s job is to keep these dependency relations valid and up-to-date. It will take actions to eventually get the registry to it's desired state.
+* Example manifest:
+  ```
+  id: "example-manifest"
+  generated_resources:
+    - pattern: apis/-/versions/-/specs/-/artifacts/lint-spectral
+      dependencies:
+        - pattern: $resource.spec
+          filter: "mime_type.contains('openapi')"
+      action: "registry compute lint $resource.spec --linter spectral"
+  ```
+  The above manifest defines that for every spec (`pattern: $resource.spec`) of type openapi (`filter: "mime_type.contains('openapi')"`) in the registry, there should exist a corresponding lint-spectral artifact (`pattern: apis/-/versions/-/specs/-/artifacts/lint-spectral`). 
+  If the target artifact is not present or out-of-date, then it can be re-generated using the specified action (`action: "registry compute lint $resource.spec --linter spectral"`).
+
+* The controller will compare the current and the supplied desired state of the registry. For resources which are non-existing or outdated, the controller will execute the action supplied in the manifest.
+
+### How to use it?
+
+With this basic definition in mind, you can supply different configurations to the controller to generate various artifacts in the registry.
+
+#### Setup the registry server:
+Make sure you have a registry server running. We will deploy the server to GKE for the purpose of this walkthrough. 
+
+- Make sure you have the right config set in [config/registry-server.yaml](https://github.com/apigee/registry/blob/main/config/registry-server.yaml).
+- From the root directory, run the following to deploy a registry server:
+  ```shell
+  # Build
+  make build
+  
+  # Deploy
+  make deploy-gke
+  ``` 
+- Populate some resources in the registry:
+  ```shell
+  # Setup auth
+  source auth/GKE.sh
+
+  # Create demo project
+  apg registry create-project --project_id demo --json
+
+  # Create 10 versions of the petstore API
+  ./tests/controller/create_apis.sh 10
+  ``` 
+
+#### Upload the "desired state" definition
```

Comment:
IMO the header here should be "Upload the manifest" with any additional explanation in the section's body. This seems like an attempt to simplify the documentation but I think it has the opposite effect here because readers don't necessarily know that "desired state definition" and "manifest" are the same thing.

If we really do want to explain again that the manifest represents a "desired state" we can link back to the first explanation. Maybe the document should have a "definitions" section explaining controller specific terminology.

Reply:
Done.